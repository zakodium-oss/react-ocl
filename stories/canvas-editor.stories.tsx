import {useCallback, useState} from 'react';

import CanvasEditor from '../src/components/CanvasEditor.js';
import type {Molecule} from "openchemlib/full";

export default {
    title: 'CanvasEditor',
    component: CanvasEditor,
    args: {
        mode: "molecule",
        fragment: false,
        width: 675,
        height: 450,
    },
    parameters: {
        docs: {
            description: {
                component: 'StructureEditor is an uncontrolled component.',
            },
        },
    },
};

const initialIDCode = 'gJX@@eKU@@ gFp@DiTt@@@!gGQHDHaImfh@##!B_vp@[G|S@AL !BmpJH@ox@?`BH?@ !Bb@K~@Hc}b@JH?P';

export function TheCanvasEditor()
{
    const [idCode, setIDCode] = useState(initialIDCode);
    const [previous, setPrevious] = useState<string | null>(null);
    const cb = useCallback(
        (molfile: string | null, molecule: Molecule | null, newIDCode: string | null) => {
            if (!newIDCode) {
                setIDCode("");
            } else {
                setIDCode(newIDCode);
            }
            setPrevious(idCode);
        },
        [setIDCode, setPrevious, idCode],
    );
    return (
        <div>
            <h2>Editor</h2>
            <CanvasEditor initialIDCode={idCode} onChange={cb} mode={"reaction"}/>
            <div style={{display: 'flex'}}>
                <div style={{flex: 1}}>
                    <h2 style={{textAlign: 'center'}}>Current</h2>
                    <pre>{idCode}</pre>
                </div>
                <div style={{flex: 1}}>
                    <h2 style={{textAlign: 'center'}}>Previous</h2>
                    <pre>{previous}</pre>
                </div>
            </div>
        </div>
    );
}
