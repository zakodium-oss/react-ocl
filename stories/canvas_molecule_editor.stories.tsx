import type { Meta, StoryObj } from '@storybook/react-vite';
import { useCallback, useState } from 'react';

import type { CanvasEditorOnChangeMolecule } from '../src/index.ts';
import { CanvasMoleculeEditor } from '../src/index.ts';

export default {
  title: 'CanvasMoleculeEditor',
  component: CanvasMoleculeEditor,
  args: {
    width: 675,
    height: 450,
    inputFormat: 'smiles',
    inputValue: 'CN1C2CCC1C(C(C2)OC(=O)C3=CC=CC=C3)C(=O)OC',
  },
} satisfies Meta<typeof CanvasMoleculeEditor>;

type Story = StoryObj<typeof CanvasMoleculeEditor>;

export const Basic: Story = {};

export const ReadOnly: Story = {
  args: {
    readOnly: true,
  },
};

export const Fragment: Story = {
  args: {
    fragment: true,
  },
};

export const Reactive: Story = {
  render(args) {
    const [data, setData] = useState({
      idcode: '',
      molfile: '',
      molfileV3: '',
      smiles: '',
    });
    const cb = useCallback((event: CanvasEditorOnChangeMolecule) => {
      setData({
        idcode: event.getIdcode(),
        molfile: event.getMolfile(),
        molfileV3: event.getMolfileV3(),
        smiles: event.getSmiles(),
      });
    }, []);
    return (
      <div>
        <h2>Reactive editor</h2>
        <p>
          The editor is not supposed to be used like this by always passing the
          state from the last event like a controlled component. This cannot
          work properly because of the way coordinates are managed internally.
          This story is here to show that it is possible to update the state by
          changing the input value, though.
        </p>
        <CanvasMoleculeEditor
          {...args}
          inputFormat="idcode"
          inputValue={data.idcode}
          onChange={cb}
        />
        <table>
          <tbody>
            <tr>
              <td valign="top">Idcode</td>
              <td>
                <code>{data.idcode}</code>
              </td>
            </tr>
            <tr>
              <td valign="top">Molfile V2000</td>
              <td>
                <code style={{ whiteSpace: 'pre' }}>{data.molfile}</code>
              </td>
            </tr>
            <tr>
              <td valign="top">Molfile V3000</td>
              <td>
                <code style={{ whiteSpace: 'pre' }}>{data.molfileV3}</code>
              </td>
            </tr>
            <tr>
              <td valign="top">SMILES</td>
              <td>
                <code>{data.smiles}</code>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    );
  },
};
