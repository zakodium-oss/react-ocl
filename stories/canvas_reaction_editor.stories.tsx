import type { Meta, StoryObj } from '@storybook/react';
import { useCallback, useState } from 'react';

import { CanvasReactionEditor } from '../src/components/canvas_editor.js';
import type { CanvasEditorOnChangeReaction } from '../src/components/canvas_editor_hook.js';

export default {
  title: 'CanvasReactionEditor',
  component: CanvasReactionEditor,
  args: {
    width: 800,
    height: 450,
    inputFormat: 'smiles',
    inputValue: 'CC(=O)O.OCC>>CC(=O)OCC',
  },
} satisfies Meta<typeof CanvasReactionEditor>;

type Story = StoryObj<typeof CanvasReactionEditor>;

export const Basic: Story = {};

export const ReadOnly: Story = {
  args: {
    readOnly: true,
  },
};

export const Reactive: Story = {
  render(args) {
    const [data, setData] = useState({
      idcode: '',
      rxn: '',
      rxnV3: '',
      smiles: '',
    });
    const cb = useCallback((event: CanvasEditorOnChangeReaction) => {
      setData({
        idcode: event.getIdcode(),
        rxn: event.getRxn(),
        rxnV3: event.getRxnV3(),
        smiles: event.getSmiles(),
      });
    }, []);
    return (
      <div>
        <h2>Reactive editor</h2>
        <p>
          The editor is not supposed to be used like this by always passing the
          state from the last event like a controlled component. This cannot
          work properly because of the way coordinates are managed internally.
          This story is here to show that it is possible to update the state by
          changing the input value, though.
        </p>
        <CanvasReactionEditor
          {...args}
          inputFormat="idcode"
          inputValue={data.idcode}
          onChange={cb}
        />
        <table>
          <tbody>
            <tr>
              <td valign="top">Idcode</td>
              <code>{data.idcode}</code>
            </tr>
            <tr>
              <td valign="top">RXN V2000</td>
              <code style={{ whiteSpace: 'pre' }}>{data.rxn}</code>
            </tr>
            <tr>
              <td valign="top">RXN V3000</td>
              <code style={{ whiteSpace: 'pre' }}>{data.rxnV3}</code>
            </tr>
            <tr>
              <td valign="top">SMILES</td>
              <td>
                <code>{data.smiles}</code>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    );
  },
};
