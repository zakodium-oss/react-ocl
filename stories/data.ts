import OCL from 'openchemlib/minimal';

export const molfileV2000 = `
  Marvin  07260611242D          

 24 26  0  0  0  0            999 V2000
   -0.6312    0.1434    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.1937    0.1434    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.5245    1.5927    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2250    0.5897    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2250    1.4147    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2187    1.9507    0.0000 C   0  0  1  0  0  0  0  0  0  0  0  0
   -0.9620    1.5927    0.0000 C   0  0  2  0  0  0  0  0  0  0  0  0
   -1.6765    2.8302    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -1.6765    2.0052    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.3910    1.5927    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -3.1055    2.0052    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2187    2.7757    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.4957    4.0132    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.4957    3.1882    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2102    2.7757    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.9247    1.5382    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.6391    1.9507    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.6391    2.7757    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.9247    3.1882    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2102    1.9507    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.1456    0.7885    0.0000 C   0  0  1  0  0  0  0  0  0  0  0  0
   -1.9425    0.5750    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.7081    0.7885    0.0000 C   0  0  1  0  0  0  0  0  0  0  0  0
    1.5050    0.5750    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  6  7  1  0  0  0  0
  3  6  1  0  0  0  0
  7 21  1  0  0  0  0
 21  1  1  0  0  0  0
  1  2  1  0  0  0  0
  2 23  1  0  0  0  0
 23  3  1  0  0  0  0
 21  4  1  1  0  0  0
 23  4  1  1  0  0  0
  4  5  1  0  0  0  0
  6 12  1  1  0  0  0
  7  9  1  1  0  0  0
  9  8  2  0  0  0  0
  9 10  1  0  0  0  0
 10 11  1  0  0  0  0
 12 14  1  0  0  0  0
 14 13  2  0  0  0  0
 14 15  1  0  0  0  0
 19 15  2  0  0  0  0
 15 20  1  0  0  0  0
 16 17  1  0  0  0  0
 20 16  2  0  0  0  0
 17 18  2  0  0  0  0
 18 19  1  0  0  0  0
 21 22  1  1  0  0  0
 23 24  1  1  0  0  0
M  END
`;

const mol = OCL.Molecule.fromMolfile(molfileV2000);

export const idcode = mol.getIDCodeAndCoordinates();

export const molfileV3000 = mol.toMolfileV3();
