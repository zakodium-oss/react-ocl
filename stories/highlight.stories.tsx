import { useState } from 'react';

import { MolfileSvgRenderer } from '../src/index.js';

import { molfileV2000 } from './data.js';

export default {
  title: 'Highlighting',
  args: {
    molfile: molfileV2000,
    atomHighlightColor: 'yellow',
    atomHighlightOpacity: 0.5,
    bondHighlightColor: 'red',
    bondHighlightOpacity: 0.5,
  },
  argTypes: {
    atomHighlightColor: {
      name: 'Atom highlight color',
      control: 'color',
    },
    atomHighlightOpacity: {
      name: 'Atom highlight opacity',
      control: 'number',
    },
    bondHighlightColor: {
      name: 'Bond highlight color',
      control: 'color',
    },
    bondHighlightOpacity: {
      name: 'Bond highlight opacity',
      control: 'number',
    },
  },
};

export function Fixed(args: any) {
  return <MolfileSvgRenderer {...args} />;
}
Fixed.storyName = 'Fixed highlight';
Fixed.args = {
  atomHighlight: [1, 5],
  bondHighlight: [6],
};
Fixed.argTypes = {
  atomHighlight: {
    name: 'Atom highlight ids',
    control: 'object',
  },
  bondHighlight: {
    name: 'Bond highlight ids',
    control: 'object',
  },
};

export function Hover(args: any) {
  const [currentAtom, setCurrentAtom] = useState(null);
  const [currentBond, setCurrentBond] = useState(null);
  return (
    <MolfileSvgRenderer
      {...args}
      atomHighlight={currentAtom ? [currentAtom] : null}
      onAtomEnter={setCurrentAtom}
      onAtomLeave={() => setCurrentAtom(null)}
      bondHighlight={currentBond ? [currentBond] : null}
      onBondEnter={setCurrentBond}
      onBondLeave={() => setCurrentBond(null)}
    />
  );
}
Hover.storyName = 'Highlight hovered element';
Hover.args = {};
