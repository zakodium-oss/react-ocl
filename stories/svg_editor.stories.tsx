import type { Meta, StoryObj } from '@storybook/react-vite';
import { useState } from 'react';

import { SvgEditor } from '../src/index.js';

import { molecule as defaultBaseMolecule } from './data.js';

const defaultMolecule = defaultBaseMolecule.getCompactCopy();
defaultMolecule.setAtomCustomLabel(0, 'Custom');

export default {
  title: 'SvgEditor',
  component: SvgEditor,
  // We cannot pass a Molecule as arg because it is not serializable.
  render: (args) => {
    const [molecule, setMolecule] = useState(defaultMolecule);

    return <SvgEditor {...args} molecule={molecule} onChange={setMolecule} />;
  },
} satisfies Meta<typeof SvgEditor>;

type Story = StoryObj<typeof SvgEditor>;

export const Control: Story = {};
