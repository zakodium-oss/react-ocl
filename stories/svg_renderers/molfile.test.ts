import { composeStories } from '@storybook/react-vite';
import { expect, test } from 'vitest';

import * as stories from './molfile.stories.tsx';

const { V2000 } = composeStories(stories);

test('Molecule renders molfile with default id', async () => {
  await V2000.run();

  expect(document.body.firstChild).toMatchSnapshot();
});
