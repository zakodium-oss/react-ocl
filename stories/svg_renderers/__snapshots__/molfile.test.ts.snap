// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Molecule renders molfile with default id 1`] = `
<div>
  <svg
    height="163px"
    id="react-ocl_r_0_"
    style="user-select: none;"
    version="1.1"
    viewBox="45 15 210 163"
    width="210px"
    xmlns="http://www.w3.org/2000/svg"
  >
    
  
    <style>
       text {font-family: sans-serif;} #react-ocl_r_0_ { pointer-events:none; } #react-ocl_r_0_ .event  { pointer-events:all; } line { stroke-linecap:round; } polygon { stroke-linejoin:round; } 
    </style>
    
  
    <text
      fill="rgb(48,80,248)"
      font-size="14"
      stroke="none"
      x="145.18"
      y="147.48"
    >
      N
    </text>
    
  
    <text
      fill="rgb(255,13,13)"
      font-size="14"
      stroke="none"
      x="103.04"
      y="83.04"
    >
      O
    </text>
    
  
    <text
      fill="rgb(255,13,13)"
      font-size="14"
      stroke="none"
      x="82.49"
      y="118.63"
    >
      O
    </text>
    
  
    <text
      fill="rgb(255,13,13)"
      font-size="14"
      stroke="none"
      x="144.97"
      y="84.61"
    >
      O
    </text>
    
  
    <text
      fill="rgb(255,13,13)"
      font-size="14"
      stroke="none"
      x="165.52"
      y="49.02"
    >
      O
    </text>
    
  
    <text
      fill="rgb(0,0,0)"
      font-size="14"
      stroke="none"
      x="95.78"
      y="147.9"
    >
      H
    </text>
    
  
    <text
      fill="rgb(0,0,0)"
      font-size="14"
      stroke="none"
      x="194.94"
      y="147.9"
    >
      H
    </text>
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="110.29"
      x2="110.29"
      y1="103.14"
      y2="85.37"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="106.69"
      x2="106.69"
      y1="103.14"
      y2="85.37"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="172.76"
      x2="172.76"
      y1="69.12"
      y2="51.35"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="169.16"
      x2="169.16"
      y1="69.12"
      y2="51.35"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="212.06"
      x2="191.51"
      y1="68.08"
      y2="79.94"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="211.65"
      x2="195.53"
      y1="72.48"
      y2="81.78"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="191.51"
      x2="212.06"
      y1="103.67"
      y2="115.53"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="195.53"
      x2="211.65"
      y1="101.83"
      y2="111.14"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="232.61"
      x2="232.61"
      y1="103.67"
      y2="79.94"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="229.01"
      x2="229.01"
      y1="101.11"
      y2="82.5"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="150.42"
      x2="129.04"
      y1="103.67"
      y2="113.97"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="171.79"
      x2="150.42"
      y1="113.97"
      y2="103.67"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="129.04"
      x2="123.76"
      y1="113.97"
      y2="137.1"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="123.76"
      x2="138.55"
      y1="137.1"
      y2="155.65"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="138.55"
      x2="162.28"
      y1="155.65"
      y2="155.65"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="162.28"
      x2="177.07"
      y1="155.65"
      y2="137.1"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="177.07"
      x2="171.79"
      y1="137.1"
      y2="113.97"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="124,137 134,140 134,138 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="134,138 134,140 144,144 145,139 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="177,137 166,138 167,140 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="167,140 166,138 156,139 157,144 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="150.24"
      x2="150.24"
      y1="135.81"
      y2="119.09"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="150,104 151,95 149,95 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="149,95 151,95 152,87 149,87 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="129,114 119,107 118,109 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="118,109 119,107 110,100 107,104 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="108.49"
      x2="94.38"
      y1="102.1"
      y2="110.25"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="81.49"
      x2="67.39"
      y1="110.25"
      y2="102.1"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="156.86"
      x2="170.96"
      y1="76.22"
      y2="68.08"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="170.96"
      x2="191.51"
      y1="68.08"
      y2="79.94"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="191.51"
      x2="191.51"
      y1="79.94"
      y2="103.67"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="212.06"
      x2="232.61"
      y1="115.53"
      y2="103.67"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="232.61"
      x2="212.06"
      y1="79.94"
      y2="68.08"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="124,137 115,138 116,140 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="116,140 115,138 106,140 107,143 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="177,137 185,140 186,138 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <polygon
      fill="rgb(0,0,0)"
      points="186,138 185,140 193,143 194,140 "
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:12"
      opacity="0"
      stroke-width="8"
      x1="108.49"
      x2="108.49"
      y1="102.1"
      y2="78.37"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:16"
      opacity="0"
      stroke-width="8"
      x1="170.96"
      x2="170.96"
      y1="68.08"
      y2="44.35"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:18"
      opacity="0"
      stroke-width="8"
      x1="212.06"
      x2="191.51"
      y1="68.08"
      y2="79.94"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:21"
      opacity="0"
      stroke-width="8"
      x1="191.51"
      x2="212.06"
      y1="103.67"
      y2="115.53"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:22"
      opacity="0"
      stroke-width="8"
      x1="232.61"
      x2="232.61"
      y1="103.67"
      y2="79.94"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:0"
      opacity="0"
      stroke-width="8"
      x1="150.42"
      x2="129.04"
      y1="103.67"
      y2="113.97"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:1"
      opacity="0"
      stroke-width="8"
      x1="171.79"
      x2="150.42"
      y1="113.97"
      y2="103.67"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:2"
      opacity="0"
      stroke-width="8"
      x1="129.04"
      x2="123.76"
      y1="113.97"
      y2="137.1"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:3"
      opacity="0"
      stroke-width="8"
      x1="123.76"
      x2="138.55"
      y1="137.1"
      y2="155.65"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:4"
      opacity="0"
      stroke-width="8"
      x1="138.55"
      x2="162.28"
      y1="155.65"
      y2="155.65"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:5"
      opacity="0"
      stroke-width="8"
      x1="162.28"
      x2="177.07"
      y1="155.65"
      y2="137.1"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:6"
      opacity="0"
      stroke-width="8"
      x1="177.07"
      x2="171.79"
      y1="137.1"
      y2="113.97"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:7"
      opacity="0"
      stroke-width="8"
      x1="123.76"
      x2="150.24"
      y1="137.1"
      y2="142.81"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:8"
      opacity="0"
      stroke-width="8"
      x1="177.07"
      x2="150.24"
      y1="137.1"
      y2="142.81"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:9"
      opacity="0"
      stroke-width="8"
      x1="150.24"
      x2="150.24"
      y1="142.81"
      y2="119.09"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:10"
      opacity="0"
      stroke-width="8"
      x1="150.42"
      x2="150.42"
      y1="103.67"
      y2="79.94"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:11"
      opacity="0"
      stroke-width="8"
      x1="129.04"
      x2="108.49"
      y1="113.97"
      y2="102.1"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:13"
      opacity="0"
      stroke-width="8"
      x1="108.49"
      x2="87.94"
      y1="102.1"
      y2="113.97"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:14"
      opacity="0"
      stroke-width="8"
      x1="87.94"
      x2="67.39"
      y1="113.97"
      y2="102.1"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:15"
      opacity="0"
      stroke-width="8"
      x1="150.42"
      x2="170.96"
      y1="79.94"
      y2="68.08"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:17"
      opacity="0"
      stroke-width="8"
      x1="170.96"
      x2="191.51"
      y1="68.08"
      y2="79.94"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:19"
      opacity="0"
      stroke-width="8"
      x1="191.51"
      x2="191.51"
      y1="79.94"
      y2="103.67"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:20"
      opacity="0"
      stroke-width="8"
      x1="212.06"
      x2="232.61"
      y1="115.53"
      y2="103.67"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:23"
      opacity="0"
      stroke-width="8"
      x1="232.61"
      x2="212.06"
      y1="79.94"
      y2="68.08"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:24"
      opacity="0"
      stroke-width="8"
      x1="123.76"
      x2="100.84"
      y1="137.1"
      y2="143.24"
    />
    
  
    <line
      class="event"
      id="react-ocl_r_0_:Bond:25"
      opacity="0"
      stroke-width="8"
      x1="177.07"
      x2="199.99"
      y1="137.1"
      y2="143.24"
    />
    
  
    <circle
      class="event"
      cx="138.55"
      cy="155.65"
      id="react-ocl_r_0_:Atom:0"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="162.28"
      cy="155.65"
      id="react-ocl_r_0_:Atom:1"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="171.79"
      cy="113.97"
      id="react-ocl_r_0_:Atom:2"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="150.24"
      cy="142.81"
      id="react-ocl_r_0_:Atom:3"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="150.24"
      cy="119.09"
      id="react-ocl_r_0_:Atom:4"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="150.42"
      cy="103.67"
      id="react-ocl_r_0_:Atom:5"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="129.04"
      cy="113.97"
      id="react-ocl_r_0_:Atom:6"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="108.49"
      cy="78.37"
      id="react-ocl_r_0_:Atom:7"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="108.49"
      cy="102.1"
      id="react-ocl_r_0_:Atom:8"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="87.94"
      cy="113.97"
      id="react-ocl_r_0_:Atom:9"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="67.39"
      cy="102.1"
      id="react-ocl_r_0_:Atom:10"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="150.42"
      cy="79.94"
      id="react-ocl_r_0_:Atom:11"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="170.96"
      cy="44.35"
      id="react-ocl_r_0_:Atom:12"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="170.96"
      cy="68.08"
      id="react-ocl_r_0_:Atom:13"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="191.51"
      cy="79.94"
      id="react-ocl_r_0_:Atom:14"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="212.06"
      cy="115.53"
      id="react-ocl_r_0_:Atom:15"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="232.61"
      cy="103.67"
      id="react-ocl_r_0_:Atom:16"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="232.61"
      cy="79.94"
      id="react-ocl_r_0_:Atom:17"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="212.06"
      cy="68.08"
      id="react-ocl_r_0_:Atom:18"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="191.51"
      cy="103.67"
      id="react-ocl_r_0_:Atom:19"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="123.76"
      cy="137.1"
      id="react-ocl_r_0_:Atom:20"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="177.07"
      cy="137.1"
      id="react-ocl_r_0_:Atom:21"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="100.84"
      cy="143.24"
      id="react-ocl_r_0_:Atom:22"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="199.99"
      cy="143.24"
      id="react-ocl_r_0_:Atom:23"
      opacity="0"
      r="8"
    />
    

    <text
      fill="rgb(127,127,127)"
      font-family="Arial, Helvetica, sans-serif"
      font-size="14 "
      text-anchor="middle"
      x="150 "
      y="173.33333333333334 "
    >
      Label p1
    </text>
  </svg>
</div>
`;
